site/app/libraries/database/DatabaseQueries.php

```diff
@@ -2909,4 +2910,226 @@ public function updateGradeableNew(\app\models\gradeable\Gradeable $gradeable) {
+        $params = [
+            $ta_graded_gradeable->getGradedGradeable()->getGradeable()->getId(),
+            $is_team ? null : $submitter_id,
+            $is_team ? $submitter_id : null,
+            $ta_graded_gradeable->getOverallComment(),
+            $ta_graded_gradeable->getUserViewedDate() !== null ?
+                DateUtils::dateTimeToString($ta_graded_gradeable->getUserViewedDate()) : null,
+        ];
+        $query = "
+            INSERT INTO gradeable_data (
+                g_id,
+                gd_user_id,
+                gd_team_id,
+                gd_overall_comment,
+                gd_user_viewed_date)
+            VALUES(?, ?, ?, ?, ?)";
+        $this->course_db->query($query, $params);
+
+        // Setup the graded gradeable with its new id
+        $ta_graded_gradeable->setIdFromDatabase($this->course_db->getLastInsertId());
+
+        // Also be sure to save the components
+        $this->updateGradedComponents($ta_graded_gradeable);
+    }
+
+    /**
+     * Updates an existing Ta Grade in the database along with its graded components/marks
+     * @param TaGradedGradeable $ta_graded_gradeable
+     */
+    private function updateTaGradedGradeable(TaGradedGradeable $ta_graded_gradeable) {
+        // If the grade has been modified, then update its properties
+        if ($ta_graded_gradeable->isModified()) {
+            $params = [
+                $ta_graded_gradeable->getOverallComment(),
+                $ta_graded_gradeable->getUserViewedDate() !== null ?
+                    DateUtils::dateTimeToString($ta_graded_gradeable->getUserViewedDate()) : null,
+                $ta_graded_gradeable->getId()
+            ];
+            $query = "
+                UPDATE gradeable_data SET 
+                  gd_overall_comment=?,
+                  gd_user_viewed_date=?
+                WHERE gd_id=?";
+            $this->course_db->query($query, $params);
+        }
+
+        // Also be sure to save the components
+        $this->updateGradedComponents($ta_graded_gradeable);
+    }
+
+    /**
+     * Creates a Ta Grade in the database if it doesn't exist, otherwise it just updates it
+     * @param TaGradedGradeable $ta_graded_gradeable
+     */
+    public function saveTaGradedGradeable(TaGradedGradeable $ta_graded_gradeable) {
+        // Ta Grades are initialized to have an id of 0 if not loaded from the db, so use that to check
+        if($ta_graded_gradeable->getId() < 1) {
+            $this->createTaGradedGradeable($ta_graded_gradeable);
+        } else {
+            $this->updateTaGradedGradeable($ta_graded_gradeable);
+        }
+    }
+
+    /**
+     * Deletes an entry from the gradeable_data table with the provided data id
+     * @param int $ta_graded_gradeable_id
+     */
+    public function deleteTaGradedGradeable($ta_graded_gradeable_id) {
+        $this->course_db->query("DELETE FROM gradeable_data WHERE gd_id=?", [$ta_graded_gradeable_id]);
```

Comment:
Feels like there should be a `DELETE FROM gradeable_component_data` and a `DELETE FROM gradeable_component_mark_data` in here as well

Reply:
Both of those tables have `ON DELETE CASCADE` on their foreign keys to `gradeable_data`, so I think we're ok